src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+/**
+ * File: AVIF IO
+ *
+ * Read and write AVIF images using libavif (https://github.com/AOMediaCodec/libavif) .
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+#include <math.h>
+
+#include "gd.h"
+#include "gd_errors.h"
+#include "gdhelpers.h"
+
+#ifdef HAVE_LIBAVIF
+#include "avif.h"
+#include "internal.h"
+// internal.h is only used for AVIF_MIN() and AVIF_CLAMP() macros
+
+/*
+  Define defaults for encoding images:
+    DEFAULT_CHROMA_SUBSAMPLING: 4:2:0 is commonly used for Chroma subsampling.
+    DEFAULT_MIN_QUANTIZER, DEFAULT_MAX_QUANTIZER: 
+      We need more testing to really know what quantizer settings are optimal,
+      but teams at Google have been using maximum=30 as a starting point.
+    DEFAULT_QUALITY: following gd conventions, -1 indicates the default.
+    DEFAULT_SPEED: AVIF_SPEED_DEFAULT is -1. This simply tells the AVIF encoder to use the default speed.
+*/
+
+#define DEFAULT_CHROMA_SUBSAMPLING AVIF_PIXEL_FORMAT_YUV420
+#define DEFAULT_QUANTIZER 30
+#define DEFAULT_QUALITY -1
+#define DEFAULT_SPEED AVIF_SPEED_DEFAULT
+
+// This initial size for the gdIOCtx is standard among GD image conversion functions
+#define NEW_DYNAMIC_CTX_SIZE 2048
+
+// Our quality param ranges from 0 to 100.
+// To calculate quality, we convert from AVIF's quantizer scale, which runs from 63 to 0.
+#define MAX_QUALITY 100
+
+// For computing the number of tiles and threads to use during encoding
+// Maximum threads are from libavif/contrib/gkd-pixbuf/loader.c
+#define MIN_TILE_AREA (512 * 512)
+#define MAX_TILES 6
+#define MAX_THREADS 64
+
+
+/*** Helper function signatures ***/
+
+static avifBool _gdImageAvifCtx (gdImagePtr im, gdIOCtx * outfile, int quality, int speed);
+static void destroyCtxAndAvifIO(avifIO *io);
+static avifIO *createAvifIOFromCtx(gdIOCtx * ctx);
+static avifResult readFromCtx(avifIO * io, uint32_t readFlags, uint64_t offset, size_t size, avifROData * out);
+static void destroyCtxAndAvifIO(avifIO *io);
+static avifBool isAvifError(avifResult result, char * msg);
+static int quality2Quantizer(int quality);
+static uint8_t convertTo8BitAlpha(uint8_t originalAlpha);
+static void setEncoderTilesAndThreads(avifEncoder * encoder, avifRGBImage * rgb);
+
+
+/*****************************************************************************
```

Comment:
this file isn't big, so you can omit these giant comment blocks

Reply:
removed